### PR TITLE
Create unsafe code section - include information on function pointers

### LINKED ISSUE #20581: Language reference : function pointer
- The `delegate` page should include `delegate*`.
- The Programming Guide section on unsafe code should add a discussion of  *function pointers* in the article on pointer types.
- If time, the unsafe code section should move from the programming guide to the language reference.

### PR DESCRIPTION
**Pull request by BillWagner**
This PR fixes #20581.

To put that new information in context, other information needs to move. Our existing content has most of the unsafe code section in the programming guide section. 

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code/?branch=pr-en-us-23559)

First commit: Move the remaining unsafe code items from the programming guide to the language reference. This makes a new section for unsafe code. None of it was in the 

Second commit: Move snippets to the right location, combine unsafe code content into one article.

### CODE REVIEW

**Comment by BillWagner** [comment]
closing and opening for a fresh build

File: docs/csharp/language-reference/unsafe-code.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+Size of the following `struct` doesn't depend on the number of elements in the array, since `pathName` is a reference:
+
+:::code language="csharp" source="snippets/unsafe-code/FixedKeywordExamples.cs" ID="6":::
+
+A `struct` can contain an embedded array in unsafe code. In the following example, the `fixedBuffer` array has a fixed size. You use a `fixed` statement to establish a pointer to the first element. You access the elements of the array through this pointer. The `fixed` statement pins the `fixedBuffer` instance field to a specific location in memory.
+
+:::code language="csharp" source="snippets/unsafe-code/FixedKeywordExamples.cs" ID="7":::
+
+The size of the 128 element `char` array is 256 bytes. Fixed size [char](builtin-types/char.md) buffers always take 2 bytes per character, regardless of the encoding. This size is true even when char buffers are marshaled to API methods or structs with `CharSet = CharSet.Auto` or `CharSet = CharSet.Ansi`. For more information, see <xref:System.Runtime.InteropServices.CharSet>.
```
Comment (gewarren) [inline docs/csharp/language-reference/unsafe-code.md:117]:
```suggestion
The size of the 128 element `char` array is 256 bytes. Fixed size [char](builtin-types/char.md) buffers always take 2 bytes per character, regardless of the encoding. This array size is the same even when char buffers are marshaled to API methods or structs with `CharSet = CharSet.Auto` or `CharSet = CharSet.Ansi`. For more information, see <xref:System.Runtime.InteropServices.CharSet>.
```

File: docs/csharp/language-reference/unsafe-code.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+Fixed size buffers are compiled with the <xref:System.Runtime.CompilerServices.UnsafeValueTypeAttribute?displayProperty=nameWithType>, which instructs the common language runtime (CLR) that a type contains an unmanaged array that can potentially overflow. Memory allocated using [stackalloc](operators/stackalloc.md) also automatically enables buffer overrun detection features in the CLR. The previous example shows how a fixed size buffer could exist in an `unsafe struct`.
+
+```csharp
+internal unsafe struct Buffer
+{
+    public fixed char fixedBuffer[128];
+}
+```
+
+The compiler-generated C# for `Buffer`, is attributed as follows:
```
Comment (gewarren) [inline docs/csharp/language-reference/unsafe-code.md:132]:
```suggestion
The compiler-generated C# for `Buffer` is attributed as follows:
```

File: docs/csharp/language-reference/unsafe-code.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+- Methods that take a `delegate*` (or return a `delegate*`) can only be called in an `unsafe` context.
+- The `&` operator to obtain the address of a function is allowed only on `static` functions. (This rule applies to both member functions and local functions).
+
+The syntax has parallels with declaring `delegate` types and using pointers. The `*` suffix on `delegate` indicates the declaration is a *function pointer*. The `&` when assigning a method group to a function pointer indicates the operation takes the address of the method.
+
+You can specify the calling convention for a `delegate*` using the keywords `managed` and `unmanaged`. In addition, for `unmanaged` function pointers, you can specify the calling convention. The following declarations show examples of each. The first uses the `managed` calling convention, which is the default. The next three use an `unmanaged` calling convention. Each specifies one of the ECMA 335 calling conventions: `Cdecl`, `Stdcall`, `Fastcall`, and `Thiscall`. The last uses the `unmanaged` calling convention, instructing the CLR to pick the default calling convention for the platform. The CLR will choose the calling convention at runtime.
```
Comment (gewarren) [inline docs/csharp/language-reference/unsafe-code.md:187]:
```suggestion
You can specify the calling convention for a `delegate*` using the keywords `managed` and `unmanaged`. In addition, for `unmanaged` function pointers, you can specify the calling convention. The following declarations show examples of each. The first declaration uses the `managed` calling convention, which is the default. The next three use an `unmanaged` calling convention. Each specifies one of the ECMA 335 calling conventions: `Cdecl`, `Stdcall`, `Fastcall`, or `Thiscall`. The last declarations uses the `unmanaged` calling convention, instructing the CLR to pick the default calling convention for the platform. The CLR will choose the calling convention at run time.
```

**Review by gewarren — COMMENTED** [review]
There are a number of site-relative links to #function-pointers. This could indicate a redirect issue.

File: docs/csharp/language-reference/keywords/fixed-statement.md [hunk 4]
```diff
@@ -69,5 +69,5 @@ For more information, see [The fixed statement](~/_csharplang/spec/unsafe-code.m
 - [C# Programming Guide](../../programming-guide/index.md)
 - [C# Keywords](index.md)
 - [unsafe](unsafe.md)
-- [Pointer types](../../programming-guide/unsafe-code-pointers/pointer-types.md)
-- [Fixed Size Buffers](../../programming-guide/unsafe-code-pointers/fixed-size-buffers.md)
+- [Pointer types](/dotnet/csharp/language-reference/unsafe-code/#pointer-types)
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/keywords/fixed-statement.md:72]: Great catch. I see the mistake there. I'll fix it and re-run the tools.

**Comment by BillWagner** [comment]
Thanks @gewarren 

I addressed all your comments.

**Review by gewarren — APPROVED** [review]
LGTM

**Pull request merged**